src/app/modules/spcp/spcp.service.ts

```diff
@@ -204,4 +212,151 @@ export class SpcpService {
+        : this.#spcpProps.spCookieMaxAge
+    }
+    if (isValidAuthenticationQuery(samlArt, destination, idpId)) {
+      return ok({
+        formId: extractFormId(destination),
+        destination,
+        rememberMe,
+        cookieDuration,
+        // Resolve known express req.query issue where pluses become spaces
+        samlArt: String(samlArt).replace(/ /g, '+'),
+      })
+    } else {
+      logger.error({
+        message: 'Invalid authentication query',
+        meta: logMeta,
+      })
+      return err(new InvalidOOBParamsError())
+    }
+  }
+
+  getSpcpAttributes(
+    samlArt: string,
+    destination: string,
+    authType: AuthType.SP | AuthType.CP,
+  ): ResultAsync<Record<string, unknown>, RetrieveAttributesError> {
+    const logMeta = {
+      action: 'getSpcpAttributes',
+      authType,
+      destination,
+      samlArt,
+    }
+    const authClient = this.getAuthClient(authType)
+    return ResultAsync.fromPromise(
+      getAttributesPromise(authClient, samlArt, destination),
+      (error) => {
+        logger.error({
+          message: 'Failed to retrieve attributes from SP/CP',
+          meta: logMeta,
+          error,
+        })
+        return new RetrieveAttributesError()
+      },
+    )
+  }
+
+  createJWT(
+    payload: JwtPayload,
+    cookieDuration: number,
+    authType: AuthType.SP | AuthType.CP,
+  ): Result<string, ApplicationError> {
+    const authClient = this.getAuthClient(authType)
+    return ok(
+      authClient.createJWT(
+        payload,
+        cookieDuration / 1000,
+        // NOTE: cookieDuration is interpreted as a seconds count if numeric.
+      ),
+    )
+  }
+
+  addLogin(
+    form: IPopulatedForm,
+    authType: AuthType.SP | AuthType.CP,
+  ): ResultAsync<ILoginSchema, AuthTypeMismatchError | DatabaseError> {
+    const logMeta = {
+      action: 'addLogin',
+      formId: form._id,
+    }
+    if (form.authType !== authType) {
+      logger.error({
+        message: 'Form auth type did not match attempted auth type',
+        meta: {
+          ...logMeta,
+          attemptedAuthType: authType,
+          formAuthType: form.authType,
+        },
+      })
+      return errAsync(new AuthTypeMismatchError(authType, form.authType))
+    }
+    return ResultAsync.fromPromise(
+      LoginModel.addLoginFromForm(form),
+      (error) => {
+        logger.error({
+          message: 'Error adding login to database',
+          meta: logMeta,
+          error,
+        })
+        return new DatabaseError()
+      },
+    )
+  }
+
+  createJWTPayload(
+    attributes: Record<string, unknown>,
+    rememberMe: boolean,
+    authType: AuthType.SP | AuthType.CP,
+  ): Result<JwtPayload, MissingAttributesError> {
+    if (authType === AuthType.SP) {
+      const userName = (attributes as SingpassAttributes).UserName
+      return userName && typeof userName === 'string'
+        ? ok({ userName, rememberMe })
+        : err(new MissingAttributesError())
+    } else {
```

Comment:
don't need else block since already returned... maybe we should enable [eslint#no-else-return](https://eslint.org/docs/rules/no-else-return) haha

Reply:
I usually write it this way when it's a "case" situation rather than a "return early because no need to do anything situation". am ok either way